itou/www/dashboard/forms.py

```diff
@@ -57,3 +58,29 @@ def clean(self):
             self._meta.model.clean_pole_emploi_fields(
                 self.cleaned_data["pole_emploi_id"], self.cleaned_data["lack_of_pole_emploi_id_reason"]
             )
+
+
+class EditUserEmailForm(forms.Form):
+
+    email = forms.EmailField(label=gettext_lazy("Nouvelle adresse e-mail"), required=True,)
+    email_confirmation = forms.EmailField(label=gettext_lazy("Confirmation"), required=True,)
+
+    def __init__(self, *args, **kwargs):
+        self.user_email = kwargs.pop("user_email")
+        super().__init__(*args, **kwargs)
+
+    def clean(self):
+        super().clean()
+        email = self.cleaned_data.get("email")
+        email_confirmation = self.cleaned_data.get("email_confirmation")
+        if email != email_confirmation:
+            raise ValidationError(gettext_lazy("Les deux adresses sont différentes."))
+        return self.cleaned_data
+
+    def clean_email(self):
+        email = self.cleaned_data["email"]
+        if email == self.user_email:
+            raise ValidationError(gettext_lazy("Veuillez indiquer une adresse différente de l'actuelle."))
+        if get_user_model().objects.filter(email=email):
```

Comment:
1) à ma grande surprise on a pas d'index sur `user.email` cf 

![image](https://user-images.githubusercontent.com/10533583/101141353-dbd3b100-361c-11eb-9ef2-8e5aa5d5eeee.png)

Ca m'étonne beaucoup que ça ne vienne pas out of the box du framework. Apparemment le framework met uniquement un index sur `account_emailaddress.email` donc je suppose que le matching passe toujours par cette table d'abord.

Pinaillage : on pourrait rajouter cet index sans quoi ton `filter(email=email)` est assez lourd niveau db. Cela profitera aussi à pas mal d'autres requêtes lourdes dans notre base de code.

2) un `if get_user_model().objects.filter(email=email).exists():` rapatrierait moins de données de la db, seulement un count.

3) je t'encourage à utiliser plutôt `get_user_model().email_already_exists(email)` qui a l'avantage d'ignorer la casse pour l'unicité (`filter(email__iexact=email)` => `Consider toto@toto.com and TOTO@toto.com as
        the same email.`)

Reply:
Merci pour ce retour ! Je ne savais pas que la méthode `get_user_model().email_already_exists(email)` existait. J'aurais clairement dû regarder...
Cela étant, je me demande s'il ne serait pas plus efficace de se baser sur Postgres au lieu de filtrer la base à chaque fois. Il me semble que l'ORM lève une erreur quand on essaie d'enregistrer un utilisateur dont le mail existe déjà. 
Je vais ajouter le test et faire quelques essais.